### ISSUE TITLE
grid search always return the value in the same position

### ISSUE THREAD

**Issue by biaohanli** [skorch-dev/skorch#640]
first I defined  a `NeuralNetClassifier` and the parameters are follows:
```python
params = {
    'lr': [0.1, 0.001, 0.0001,0.01],
    'batch_size': [50, 100, 150, 200, 250, 300],
}
skf = StratifiedKFold(n_splits=4, shuffle=True, random_state=1)
gs = RandomizedSearchCV(net, params, 2,
                        scoring=make_scorer(log_loss, needs_proba=True, labels=y_data), cv=skf)
gs.fit(x,y)
```

I found out that the skorch always returns the second parameter in lr and the third parameter in batch_size

does anyone know why?

**Comment by BenjaminBossan** [comment]
> I found out that the skorch always returns the second parameter in lr and the third parameter in batch_size

Can you elaborate on that? I don't understand what you mean here. Ideally, you can show some code that produces the mentioned outcome.

(Btw, after adding code on github, please select it and click on the "insert code" button (looks like this: `<>`), which makes it more readable)

**Comment by biaohanli** [comment]
the code is as follows:
class Net(nn.Module):
    def __init__(self):
        pass

    # def forward(self, x, value):
    def forward(self, x, y=None):
        pass


model = Net()
if args.cuda:
    model.cuda()

x_data, y_data = next(iter(train_loader))
print(x_data.shape)
print(y_data.shape)
net = NeuralNetClassifier(Net
                          , max_epochs=200
                          , lr=0.001
                          , batch_size=64
                          , verbose=1)

params = {
    'lr': [0.1, 0.001, 0.0001,0.01],
    'batch_size': [50, 100, 150, 200, 250, 300],
}

skf = StratifiedKFold(n_splits=4, shuffle=True, random_state=1)
gs = RandomizedSearchCV(net, params, 2,
                        scoring=make_scorer(log_loss, needs_proba=True, labels=y_data), cv=skf)
gs.fit(x_data, y_data)
print(gs.best_score_, gs.best_params_, gs.best_estimator_)

the wired thing is that , RandomizedSearchCV always return the second parameter in lr and the third parameter in batch_size.
in this case it is lr = 0.001 and batch_size=150
when I change params:
params = {
    'lr': [0.1, 0.02,0.001, 0.0001,0.01],
    'batch_size': [50, 100, 200,150, 250, 300],
}
the results changes to lr = 0.02 and batch_size=200

I also tried GridSearch without random,it is the same.

**Comment by BenjaminBossan** [comment]
I have a few questions here:

> def forward(self, x, y=None):
>     pass

Is this your actual code? Because that shouldn't even run.

> model = Net()
> if args.cuda:
> model.cuda()

This code doesn't do anything, because you're not using `model` later on. Anyway, it is better to pass the module class (in your case `Net`), not the instance, to `NeuralNetClassifier` (as you did) and pass `device='cuda'` as an argument.

> RandomizedSearchCV(net, params, 2,

Have you tested if anything changes if you use more than 2 iterations?

> make_scorer(log_loss, needs_proba=True, labels=y_data)

What do you need `labels=y_data` for? I don't think this is used at all.

> RandomizedSearchCV always return the second parameter in lr and the third parameter in batch_size

So what you mean is that the best parameter is always in the same position, no matter the value? Is it the same even if you change the random state? If that is so, and you're training on real data, it would be strange, but possibly unrelated to skorch and more an sklearn thing.

I would suggest to do the following: Use a more realistic model and enough data. First, don't use grid/randomized search at all, only skorch, and see if the skorch model trains at all and if you can get some decent validation scores (you could post the print log here). Once you've got that, you have a rough idea about where in the hyper parameter space to search. Then you can mold your `params` around those values and run grid/randomized search with more iterations.

If at that point you still always get back the same best parameters, even though skorch's `NeuralNetClassifier` seems to be learning, it might be a bug in sklearn (but I think that's very unlikely). Please also take a look at `gs.cv_results_`, check if the scores are always the same or are different for different parameters. If the scores are always exactly the same, that could be somehow the reason for what you're seeing.

**Comment by BenjaminBossan** [comment]
@biaohanli any updates?

**Comment by BenjaminBossan** [comment]
@biaohanli Since we haven't heard back in more than a month, I'll close the issue. Feel free to re-open if you have new input.
